### PR TITLE
Two Factor Authentication

### PR DESCRIPTION
**Pull request by CityVibes**
Feel free to review code, note: **DO NOT MERGE** - need to update cloudsdk to 0.5.1 (once released) for 2fa flow to work.

### CODE REVIEW

File: devicesetup/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -61,6 +61,7 @@
     <string name="verify_product_ownership">Verifying product ownership</string>
     <string name="action_done">Done</string>
     <string name="title_activity_password_reset">Password Reset</string>
+    <string name="title_activity_two_factor">Two Factor Authentication</string>
```
Comment (RaimundasSakalauskas) [inline devicesetup/src/main/res/values/strings.xml]: @channel — something we decided in our stand up today around communication:

- In _internal_ and _technical_ communications, we will refer to this feature as *multi-factor authentication (MFA)*
- In _external_ and _customer-facing_ communications, we will refer to this feature as *two-step authentication*

**Review by RaimundasSakalauskas — APPROVED** [review]
Except that one string I commented, looks good to me.

File: devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            login(callingIntent.getStringExtra(EMAIL_EXTRA), callingIntent.getStringExtra(PASSWORD_EXTRA),
+                    callingIntent.getStringExtra(MFA_EXTRA), code);
+        }
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_two_factor);
+        ParticleDeviceSetupLibrary.getInstance().getApplicationComponent().activityComponentBuilder()
```
Comment (jensck) [inline devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java]: bro, do you even line break?  ;)

Again, not blocking PR for this, but in the future, let's use builder-style wrapping for this kind of long call chain.

File: devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        TwoFactorActivity.this,
+                        sparkCloud,
+                        SDKGlobals.getSensitiveDataStorage(),
+                        null));
+                finish();
+            }
+
+            @Override
+            public void onFailure(@NonNull ParticleCloudException error) {
+                log.d("onFailed(): " + error.getMessage());
```
Comment (jensck) [inline devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java]: This should have the same `if (isFinishing())` call here as in `onSuccess()`.

**Review by jensck — CHANGES_REQUESTED** [review]
The `isFinishing()` thing in `TwoFactorActivity` needs to be fixed, everything else is optional/future reference!

File: devicesetup/src/main/java/io/particle/android/sdk/accountsetup/LoginActivity.java [hunk 4]
```diff
@@ -73,8 +74,15 @@ protected void afterInput() {
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.particle_activity_login);
-        ParticleDeviceSetupLibrary.getInstance().getApplicationComponent().activityComponentBuilder()
-                .apModule(new ApModule()).build().inject(this);
+
+        ParticleDeviceSetupLibrary
```
Comment (jensck) [inline devicesetup/src/main/java/io/particle/android/sdk/accountsetup/LoginActivity.java]: :+1: thanks!

File: devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java [hunk 5]
```diff
@@ -40,6 +43,20 @@
      */
     private Async.AsyncApiWorker<ParticleCloud, Void> loginTask = null;
 
+    public static Intent buildIntent(Context context, String email, String password, String mfa) {
```
Comment (jensck) [inline devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java]: +1 for adding this.

**Review by jensck — APPROVED** [review]

**Comment by jensck** [comment]
LGTM

**Pull request merged**